Skill DB TXT files to LibConfig + converter

Smokexyz opened the pull request:
This is a request to merge Skill DB TXT Files to Libconfig. It will read and convert from the following files -
```
skill_require_db.txt
skill_unit_db.txt
skill_castnodex_db.txt
skill_cast_db.txt
skill_db.txt
```

into one skill_db.conf. A php CLI converter is also included for the conversion of both re/pre-re folders.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/herculesws/hercules/1333)
<!-- Reviewable:end -->

MishimaHaruna commented:
Reviewed 3 of 6 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgHp6AFWA6AUNqOmeJ:-KLgHp6AFWA6AUNqOmeK:1037961257) (no related file):*
Pending rebase.

When rebasing, please split the HPM-related changes to a separate commit (or just omit them, and the API bot will automatically do it in a separate commit), and squash the rest into one commit.

For a better result, you can split it in three (four) commits:
- The converter script
- The db conversion
- The source changes
- (the HPM)

+codereview:needsedits 

---

*[src/map/skill.c, line 18983 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgCC8mLnQqsuhAfg40:-KLgCC8mLnQqsuhAfg41:868111592) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L18983)):*
This could leave entries not initialized in `arr`. Is it intentional? (same thing in the case below)


> (reply)
> In the event that the config isn't found, the functions setting_lookup_int and setting_get_int_elem return 0, therefore also initializing the array element to 0.


The documentation of the family of functions `config_setting_lookup_XXX()` says:

> These functions look up the value of the child setting named `name` of the setting setting. They store the value at `value` and return `CONFIG_TRUE` on success. If the setting was not found or if the type of the value did not match the type requested, they leave the data pointed to by value unmodified and return `CONFIG_FALSE`.

The array entries are therefore uninitialized if the lookup fails.

---

*[src/map/skill.c, line 19020 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgD95QXzTmDad5UwkJ:-KLgD95QXzTmDad5UwkK:1776468190) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L19020)):*
> ```C
> int skill_validate_skilltype(const char *name, bool type, struct s_skill_db *sk)
> {
> 	if (type) {
> ```

Wouldn't it be better to nest this the other way around?

```
if (strcmp(name, "foo") == 0) {
    if (type)
        sk->inf |= INF_FOO;
    else
        sk->inf &= ~INF_FOO;
} else if (strcmp(name, "bar") == 0) {
    if (type)
        sk->inf |= INF_BAR;
    else
        sk->inf &= ~INF_BAR;
}
```

That would make it easier to add new values, without forgetting to add them to one of the two sides (you edit only in one place).

---

*[src/map/skill.c, line 20232 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgFSIOiLFi9JbsDM_e:-KLgFSIOiLFi9JbsDM_f:168592108) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L20232)):*
> ```C
> 	}
> 
> #define skilldb_duplicate_warning(name, setting, skill, source) (ShowError("skill_read_skilldb: Duplicate entry '%s' in setting '%s' for Skill Id %d in '%s', skipping...\n", name, setting, skill, source))
> ```

When you #define something inside a function, please #undef it before the end of the function, when it's no longer needed (so that local macros don't clutter the global namespace)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+		/* Range */
+		if ((t=libconfig->setting_get_member(conf, "Range")))
+			skill->config_set_level(t, tmp_db.range);
+		/* Hit */
```
Hit type probably better detect in separate function. This need for plugins to override.

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+			else
+				skill->readdb_invalid_error("Hit", tstr, skill_id, filepath);
+		}
+		/* Skill Type */
```
skill type too. see previous comment

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+				}
+			}
+		}
+		/* Skill Info */
```
And into into separate function too

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+				}
+			}
+		}
+		/* Skill Attack Type */
```
and attack type. see previous comments

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+			else
+				skill->readdb_invalid_error(tstr, "AttackType", skill_id, filepath);
+		}
+		/* Skill Element */
```
and element. see previous comments

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+			else
+				skill->readdb_invalid_error(tstr, "Element", skill_id, filepath);
+		}
+		/* Damage Type */
```
damage type too. see previous comments

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+		/**
+		 * Skill Requirements data handling
+		 */
+		if ((t=libconfig->setting_get_member(conf, "Requirements")) && config_setting_is_group(t)) {
```
skill requiments too. see previous comments

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+			if ((tt = libconfig->setting_get_member(t, "AmmoAmount")))
+				skill->config_set_level(tt, tmp_db.ammo_qty);
+			/* State */
+			if (libconfig->setting_lookup_string(t, "State", &tstr) && strcmpi(tstr,"None") != ST_NONE) {
```
and state. see previous comments

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+		/**
+		 * Skill Unit data handling
+		 */
+		if ((t=libconfig->setting_get_member(conf, "Unit")) && config_setting_is_group(t)) {
```
Skill unit too. see previous comments

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+			if(libconfig->setting_lookup_int(t, "Interval", &temp))
+				tmp_db.unit_interval = temp;
+			
+			if(libconfig->setting_lookup_string(t, "Target", &tstr)) {
```
Target too. see previous comments

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1095 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+					tmp_db.unit_target |= BL_SKILL;
+			}
+			
+			if ((tt=libconfig->setting_get_member(t, "Flag")) && config_setting_is_group(tt)) {
```
and flag. see previous comments

4144 commented:
Also in all function what can be created for parse part of skill should one of parameter with tmp_db for plugins usage. for know what skill parsed now

Smokexyz commented:
Review status: 3 of 6 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[src/map/skill.c, line 19676 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-19676:-KLh01lxejvgpx_xT9vi:-835720118) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L19676)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> Hit type probably better detect in separate function. This need for plugins to override.
</details>
Done, thanks.

---

*[src/map/skill.c, line 19687 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-19687:-KLhT8dxIouD3Xq7AG31:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L19687)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> skill type too. see previous comment
</details>
Done.

---

*[src/map/skill.c, line 19704 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-19704:-KLhT9HkHUQUwyM23bU2:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L19704)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> And into into separate function too
</details>
Done.

---

*[src/map/skill.c, line 19721 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-19721:-KLhT9lbBQfORSMrUTTK:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L19721)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> and attack type. see previous comments
</details>
Done.

---

*[src/map/skill.c, line 19732 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-19732:-KLhTAMO6Mx-mvzQE1lf:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L19732)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> and element. see previous comments
</details>
Done.

---

*[src/map/skill.c, line 19859 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-19859:-KLhTAqH4hDMtksWXoVP:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L19859)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> skill requiments too. see previous comments
</details>
Done.

---

*[src/map/skill.c, line 19934 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-19934:-KLhTDCwXWwcXcnq8fO1:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L19934)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> and state. see previous comments
</details>
Done.

---

*[src/map/skill.c, line 19988 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-19988:-KLhTDeiFo9rfSAF1Tma:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L19988)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> Skill unit too. see previous comments
</details>
Done.

---

*[src/map/skill.c, line 20003 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-20003:-KLhTE6LP0mQlrVGIMv1:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L20003)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> Target too. see previous comments
</details>
Done.

---

*[src/map/skill.c, line 20029 \[r1\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r1-20029:-KLhTEUf9emKLGglQeAe:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0c3eae7fc4e58708193baad10479427bf668a303/src/map/skill.c#L20029)):*
<details><summary><i>Previously, 4144 (Andrei Karas) wrote…</i></summary>
> and flag. see previous comments
</details>
Done.

---

*[src/map/skill.c, line 18983 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgCC8mLnQqsuhAfg40:-KLh-SDFw73EzzpOnV6K:-499146009) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L18983)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> This could leave entries not initialized in `arr`. Is it intentional? (same thing in the case below)
> 
> 
> > (reply)
> > In the event that the config isn't found, the functions setting_lookup_int and setting_get_int_elem return 0, therefore also initializing the array element to 0.
> 
> 
> The documentation of the family of functions `config_setting_lookup_XXX()` says:
> 
> > These functions look up the value of the child setting named `name` of the setting setting. They store the value at `value` and return `CONFIG_TRUE` on success. If the setting was not found or if the type of the value did not match the type requested, they leave the data pointed to by value unmodified and return `CONFIG_FALSE`.
> 
> The array entries are therefore uninitialized if the lookup fails.
</details>
Done.

---

*[src/map/skill.c, line 19020 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgD95QXzTmDad5UwkJ:-KLg_kQ0FvflbSO7wcHY:978539034) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L19020)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> Wouldn't it be better to nest this the other way around?
> 
> ```
> if (strcmp(name, "foo") == 0) {
>     if (type)
>         sk->inf |= INF_FOO;
>     else
>         sk->inf &= ~INF_FOO;
> } else if (strcmp(name, "bar") == 0) {
>     if (type)
>         sk->inf |= INF_BAR;
>     else
>         sk->inf &= ~INF_BAR;
> }
> ```
> 
> That would make it easier to add new values, without forgetting to add them to one of the two sides (you edit only in one place).
</details>
Done, though this would override the previous duplicate settings if any, without a duplicate warning.

---

*[src/map/skill.c, line 20232 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgFSIOiLFi9JbsDM_e:-KLgy5d9ZV9IzPaHOzeH:1551024003) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L20232)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> When you #define something inside a function, please #undef it before the end of the function, when it's no longer needed (so that local macros don't clutter the global namespace)
</details>
Woops, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

MishimaHaruna commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/map/skill.c, line 18983 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgCC8mLnQqsuhAfg40:-KLhZFjzk9GVcdv0cEwr:-931957530) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L18983)):*
<details><summary><i>Previously, Smokexyz wrote…</i></summary>
> Done.
</details>
Umm, I don't see changes applied? Or am I missing something? o.o

---

*[src/map/skill.c, line 19020 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgD95QXzTmDad5UwkJ:-KLhXTj5SHA2_4dCJR2R:748658955) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L19020)):*
<details><summary><i>Previously, Smokexyz wrote…</i></summary>
> Done, though this would override the previous duplicate settings if any, without a duplicate warning.
</details>
That's true... But considering that the rest of our code doesn't warn on duplicates, I think it won't hurt...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

Smokexyz commented:
*[src/map/skill.c, line 18983 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgCC8mLnQqsuhAfg40:-KLireyrkKBYNvAiu9cG:-148312672) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L18983)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> Umm, I don't see changes applied? Or am I missing something? o.o
</details>
Sorry I'm not clear on what needs to be done here, I've initiliaized the struct to be { 0 }, so in`config_setting_get_int` if a setting isn't found the data remains unmodified. As for `config_setting_get_int_elem`, a 0 is returned if a setting isn't found, which is being assigned to the element of the array. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

4144 commented on src/map/skill.c:
```diff
@@ -19180,6 +18967,1259 @@ bool skill_parse_row_changematerialdb(char* split[], int columns, int current) {
+		
+		// Validate the skill entry, add it to the duplicate array and increment count on success.
+		if ((duplicate[idx] = skill->validate_skilldb(&tmp_db, filepath)))
+			count++;
```
Probably here need add call to empty function for read additional custom fields from skill_db for plugins.
See how it implemented in items_db (itemdb_readdb_additional_fields)

Smokexyz commented:
*[src/map/skill.c, line 20212 \[r7\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLdQocEg2nE-wyBaTPq-r7-20212:-KLklQl9rjPzBs8uAQ35:-336940536) ([raw file](https://github.com/herculesws/hercules/blob/49ffb80655f198e4577be720ac2c52103b0816a1/src/map/skill.c#L20212)):*
<details><summary><i>Previously, Smokexyz wrote…</i></summary>
> Done.
</details>
Locally, will wait for more input or anything else before committing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

MishimaHaruna commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/map/skill.c, line 18983 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgCC8mLnQqsuhAfg40:-KLl1mq28ZJ4QOPgBDN8:950423662) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L18983)):*
<details><summary><i>Previously, Smokexyz wrote…</i></summary>
> Sorry I'm not clear on what needs to be done here, I've initiliaized the struct to be { 0 }, so in`config_setting_get_int` if a setting isn't found the data remains unmodified. As for `config_setting_get_int_elem`, a 0 is returned if a setting isn't found, which is being assigned to the element of the array. 
</details>
Ah, I see. I didn't notice the initialization to `{ 0 }`, sorry

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

Smokexyz commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[src/map/skill.c, line 18983 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgCC8mLnQqsuhAfg40:-KLlD1jZQU-jF0vLF019:-411689456) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L18983)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> Ah, I see. I didn't notice the initialization to `{ 0 }`, sorry
</details>
np np, whenever it's good I can get on with rebasing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

MishimaHaruna commented:
Reviewed 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[src/map/skill.c, line 18983 \[r2\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgCC8mLnQqsuhAfg40:-KLpJcvharcTOnbEumz5:-2054324852) ([raw file](https://github.com/herculesws/hercules/blob/0782bf46ebd8b9faaf980847290557dc404ce954/src/map/skill.c#L18983)):*
<details><summary><i>Previously, Smokexyz wrote…</i></summary>
> np np, whenever it's good I can get on with rebasing.
</details>
Sure, please do. Once rebased, I'll start compiling/testing it, and giving it a final review pass.

Not sure if @4144 has any more input to give (but that can be done after rebasing as well).

If you need any help or recommendations about the operations of rebase, squash or commit splitting, feel free to contact me over IRC (especially considering that it's likely there will be conflicts)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

Smokexyz commented:
Rebased commit includes source changes and the converter script. If all okay, the next commit can include the skill_db.conf files and another commit for the HPM changes.

MishimaHaruna commented:
> Since my branch is not updated and there were a considerable amount of changes to HPM, It'd be wise that I let HPM slide (even though i could use the HPMHookGen tools) and let the API follow up.

Sure, you can remove the HPM changes, and let the HerculesWS API bot take care of it upon merge

---

Reviewed 3 of 3 files at r9, 2 of 2 files at r10, 2 of 2 files at r11.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[src/map/skill.c, line 20008 \[r11\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLsS2UEjHT8I7mIs9Du:-KLsS2UEjHT8I7mIs9Dv:-1406454825) ([raw file](https://github.com/herculesws/hercules/blob/8c155749d19c9a34fc1e1c4dfcbe4dcb2b968842/src/map/skill.c#L20008)):*
> ```C
> 	if ((sk=libconfig->setting_get_member(skilldb.root, "skill_db")) == NULL) {
> 		ShowError("skill_read_skilldb: Skill DB could not be loaded, please check '%s'.\n", filepath);
> 		return false;
> ```

Before returning, you need to destroy the `skilldb` object (`libconfig->destroy(&skilldb);`) to avoid leaking it

---

*[src/map/skill.c, line 20024 \[r11\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLsRGRlHgs2Tey3TP24:-KLsRGRlHgs2Tey3TP25:-396338023) ([raw file](https://github.com/herculesws/hercules/blob/8c155749d19c9a34fc1e1c4dfcbe4dcb2b968842/src/map/skill.c#L20024)):*
> ```C
> 		tmp_db.nameid = skill_id;
> 		
> 		if((idx = skill->get_index(skill_id)) == 0)
> ```

When this happens (skill ID out of range, or within a range reserved to guild/homunculus/mercenary/elemental skills), it'd be nice to show a warning instead of silently ignore the entry

---

*[src/map/skill.c, line 20226 \[r11\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLsSfY9IztYqusd-CRk:-KLsSfY9IztYqusd-CRl:343322389) ([raw file](https://github.com/herculesws/hercules/blob/8c155749d19c9a34fc1e1c4dfcbe4dcb2b968842/src/map/skill.c#L20226)):*
> ```C
> 			count++;
> 	}
> 	
> ```

Once done with the object (and in any case before returning) please destroy `skilldb` (see comment above)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333#-:-KLsNHSuV4cFV2mE0p_z:1727088725)*
<!-- Sent from Reviewable.io -->

Smokexyz commented:
@MishimaHaruna Thanks I wasn't aware of `libconfig->destroy()`. Rebased with changes mentioned, builds are now passing.

MishimaHaruna commented:
Reviewed 4 of 4 files at r12, 12 of 12 files at r13.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

MishimaHaruna commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[tools/skilldbconverter.php, line 48 \[r13\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KMAmmoppS8ABPr1uBpy:-KMAmmoppS8ABPr1uBpz:-1175983901) ([raw file](https://github.com/herculesws/hercules/blob/f90b4803748519831ebb8cd67bbabf3622eaf704/tools/skilldbconverter.php#L48)):*
> ```PHP
> $renewal =  (issetarg("-re") || issetarg("--renewal"));
> $prere = (issetarg("-pre-re") || issetarg("--pre-renewal"));
> $constants = (issetarg("--use-constants") || issetarg("-c"));
> ```

It'd be nice if --use-constants was enabled by default (and the DB we ship had it enabled)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

Smokexyz commented:
Review status: 2 of 15 files reviewed at latest revision, 13 unresolved discussions.

---

*[tools/skilldbconverter.php, line 48 \[r13\]](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KMAmmoppS8ABPr1uBpy:-KMDFcArWShhOib7-iu7:-1310602519) ([raw file](https://github.com/herculesws/hercules/blob/f90b4803748519831ebb8cd67bbabf3622eaf704/tools/skilldbconverter.php#L48)):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> It'd be nice if --use-constants was enabled by default (and the DB we ship had it enabled)
</details>
Done, changed options to -itid/--use-itemid instead and rebased.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

MishimaHaruna commented:
Reviewed 13 of 13 files at r14, 12 of 12 files at r15.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/herculesws/hercules/1333#-KLgHp6AFWA6AUNqOmeJ:-KMEnWypfi4Bhpuuux8p:8167367) (no related file):*
<details><summary><i>Previously, MishimaHaruna (Haru) wrote…</i></summary>
> Pending rebase.
> 
> When rebasing, please split the HPM-related changes to a separate commit (or just omit them, and the API bot will automatically do it in a separate commit), and squash the rest into one commit.
> 
> For a better result, you can split it in three (four) commits:
> - The converter script
> - The db conversion
> - The source changes
> - (the HPM)
> 
> +codereview:needsedits 
</details>
-codereview:needsedits 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

MishimaHaruna commented:
Reviewed 12 of 12 files at r16, 12 of 12 files at r17.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/herculesws/hercules/1333)*
<!-- Sent from Reviewable.io -->

MishimaHaruna commented:
Thank you very much, merged!

kyeme commented:
The KN_SPEARBOOMERANG skill maximum level is up to 5 only
https://github.com/HerculesWS/Hercules/blob/master/db/pre-re/skill_db.conf#L1980

hemagx commented:
@kyeme long story, short is
this values existed and was being generated in hidden by magic in the server, and so we copied the garbage to skill_db.conf. to ensure nothing get generated by this hidden magic in the server until we fix (refactor it is more correct term) the database reading.
